packages/grammar-selector/lib/grammar-list-view.js

```diff
@@ -82,32 +82,64 @@ class GrammarListView {
   async toggle () {
     if (this.panel != null) {
       this.cancel()
-    } else if (atom.workspace.getActiveTextEditor()) {
-      this.editor = atom.workspace.getActiveTextEditor()
+      return
+    }
+
+    const editor = atom.workspace.getActiveTextEditor()
+    if (editor) {
+      this.editor = editor
       this.currentGrammar = this.editor.getGrammar()
       if (this.currentGrammar === atom.grammars.nullGrammar) {
         this.currentGrammar = this.autoDetect
       }
 
-      const grammars = atom.grammars.getGrammars(true).filter((grammar) => {
+      let grammars = atom.grammars.getGrammars({includeTreeSitter: true}).filter(grammar => {
         return grammar !== atom.grammars.nullGrammar && grammar.name
       })
+
+      if (atom.config.get("grammar-selector.hideDuplicateTextMateGrammars")) {
+        const oldGrammars = grammars
+        grammars = []
+        const blacklist = new Set()
+        for (const grammar of oldGrammars) {
+          if (isTreeSitter(grammar)) {
+            blacklist.add(grammar.name)
+            grammars.push(grammar)
+          }
+        }
+        atom.grammars.getGrammars({includeTreeSitter: false}).forEach(grammar => {
+          if (grammar !== atom.grammars.nullGrammar && grammar.name && !blacklist.has(grammar.name)) {
+            grammars.push(grammar)
+          }
+        })
+      }
+
       grammars.sort((a, b) => {
         if (a.scopeName === 'text.plain') {
           return -1
         } else if (b.scopeName === 'text.plain') {
           return 1
-        } else if (a.name) {
-          return a.name.localeCompare(b.name)
-        } else if (a.scopeName) {
-          return a.scopeName.localeCompare(b.scopeName)
-        } else {
-          return 1
+        } else if (a.name === b.name) {
```

Comment:
Equivalent logic (besides TS check), as we filtered out grammars with no name already.